pkg/reconciler/inmemorychannel/inmemorychannel.go

```diff
@@ -0,0 +1,295 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inmemorychannel
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/knative/eventing/pkg/apis/messaging/v1alpha1"
+	messaginginformers "github.com/knative/eventing/pkg/client/informers/externalversions/messaging/v1alpha1"
+	listers "github.com/knative/eventing/pkg/client/listers/messaging/v1alpha1"
+	"github.com/knative/eventing/pkg/logging"
+	"github.com/knative/eventing/pkg/reconciler"
+	"github.com/knative/eventing/pkg/utils"
+	"github.com/knative/pkg/controller"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	appsv1informers "k8s.io/client-go/informers/apps/v1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	appsv1listers "k8s.io/client-go/listers/apps/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// ReconcilerName is the name of the reconciler
+	ReconcilerName = "InMemoryChannels"
+
+	// controllerAgentName is the string used by this controller to identify
+	// itself when creating events.
+	controllerAgentName = "in-memory-channel-controller"
+
+	finalizerName = controllerAgentName
+
+	// Name of the corev1.Events emitted from the reconciliation process.
+	channelReconciled         = "InMemoryChannelReconciled"
+	channelReconcileFailed    = "InMemoryChannelReconcileFailed"
+	channelUpdateStatusFailed = "InMemoryChannelUpdateStatusFailed"
+	channelDispatcherFailed   = "InMemoryChannelDispatcherDeploymentFailed"
+	channelServiceFailed      = "InMemoryChannelDispatcherServiceFailed"
+	channelEndpointsFailed    = "InMemoryChannelDispatcherEndpointsFailed"
+)
+
+type Reconciler struct {
+	*reconciler.Base
+
+	dispatcherNamespace      string
+	dispatcherDeploymentName string
+	dispatcherServiceName    string
+
+	inmemorychannelLister   listers.InMemoryChannelLister
+	inmemorychannelInformer cache.SharedIndexInformer
+	deploymentLister        appsv1listers.DeploymentLister
+	serviceLister           corev1listers.ServiceLister
+	endpointsLister         corev1listers.EndpointsLister
+	impl                    *controller.Impl
+}
+
+var deploymentGVK = appsv1.SchemeGroupVersion.WithKind("Deployment")
+var serviceGVK = corev1.SchemeGroupVersion.WithKind("Service")
+
+// Check that our Reconciler implements controller.Reconciler.
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func FilterWithNameAndNamespace(namespace, name string) func(obj interface{}) bool {
```

Comment:
why public?

Reply:
I'm thinking of moving this to knative/pkg since it's a handy filter, so then it will need to be public. Thoughts?

Reply:
Moving to pkg makes sense